File: components/gitpod-protocol/java/src/main/java/io/gitpod/gitpodprotocol/api/entities/Workspace.java
// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License.AGPL.txt in the project root for license information.

package io.gitpod.gitpodprotocol.api.entities;

public class Workspace {

    private String id;

    private String contextURL;

    private WorkspaceContext context;

    private String creationTime;

    private String type;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getContextURL() {
        return contextURL;
    }

    public void setContextURL(String contextURL) {
        this.contextURL = contextURL;
    }

    public WorkspaceContext getContext() {
        return context;
    }

    public void setContext(WorkspaceContext context) {
        this.context = context;
    }

    public String getCreationTime() {
        return creationTime;
    }

    public void setCreationTime(String creationTime) {
        this.creationTime = creationTime;
    }

    public WorkspaceType getType() {
        return WorkspaceType.valueOf(type);
    }

    public void setType(WorkspaceType type) {
        this.type = type.toString();
    }
}

File: components/server/src/workspace.rs
// Copyright (c) 2025 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License.AGPL.txt in the project root for license information.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use crate::database::{Workspace, WorkspaceStatus};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkspaceConfig {
    pub image: String,
    pub resources: ResourceConfig,
    pub environment: Vec<EnvVar>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceConfig {
    pub cpu: String,
    pub memory: String,
    pub storage: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnvVar {
    pub name: String,
    pub value: String,
}

impl Default for WorkspaceConfig {
    fn default() -> Self {
        Self {
            image: "gitpod/workspace-full".to_string(),
            resources: ResourceConfig {
                cpu: "2".to_string(),
                memory: "4Gi".to_string(),
                storage: "10Gi".to_string(),
            },
            environment: vec![
                EnvVar {
                    name: "GITPOD_WORKSPACE_ID".to_string(),
                    value: "".to_string(),
                },
            ],
        }
    }
}

pub struct WorkspaceManager {
    // In a real implementation, this would contain Kubernetes client, etc.
}

impl WorkspaceManager {
    pub fn new() -> Self {
        Self {}
    }

    pub async fn create_workspace(&self, workspace: &Workspace, config: &WorkspaceConfig) -> Result<()> {
        tracing::info!("Creating workspace {} with image {}", workspace.id, config.image);

        // Simulate workspace creation
        tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;

        Ok(())
    }

    pub async fn start_workspace(&self, workspace_id: Uuid) -> Result<String> {
        tracing::info!("Starting workspace {}", workspace_id);

        // Simulate workspace startup
        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;

        let url = format!("https://{}.gitpod.example.com", workspace_id);
        Ok(url)
    }

    pub async fn stop_workspace(&self, workspace_id: Uuid) -> Result<()> {
        tracing::info!("Stopping workspace {}", workspace_id);

        // Simulate workspace shutdown
        tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

        Ok(())
    }

    pub async fn delete_workspace(&self, workspace_id: Uuid) -> Result<()> {
        tracing::info!("Deleting workspace {}", workspace_id);

        // Simulate workspace deletion
        tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

        Ok(())
    }

    pub async fn get_workspace_logs(&self, workspace_id: Uuid) -> Result<Vec<String>> {
        tracing::info!("Getting logs for workspace {}", workspace_id);

        // Simulate log retrieval
        Ok(vec![
            format!("[{}] Workspace starting", chrono::Utc::now().format("%Y-%m-%d %H:%M:%S")),
            format!("[{}] Installing dependencies", chrono::Utc::now().format("%Y-%m-%d %H:%M:%S")),
            format!("[{}] Workspace ready", chrono::Utc::now().format("%Y-%m-%d %H:%M:%S")),
        ])
    }
}
